pkg/kubelet/images/puller.go

```diff
@@ -34,42 +36,109 @@ type pullResult struct {
 }
 
 type imagePuller interface {
-	pullImage(context.Context, kubecontainer.ImageSpec, []v1.Secret, chan<- pullResult, *runtimeapi.PodSandboxConfig)
+	pullImage(context.Context, kubecontainer.ImageSpec, []v1.Secret, chan pullResult, *runtimeapi.PodSandboxConfig)
 }
 
 var _, _ imagePuller = &parallelImagePuller{}, &serialImagePuller{}
 
 type parallelImagePuller struct {
 	imageService kubecontainer.ImageService
 	tokens       chan struct{}
+	pulls        sync.Map
+}
+
+type pullBroker struct {
+	stopCh      chan struct{}
+	publishCh   chan pullResult
+	subscribeCh chan chan pullResult
+}
+
+func newBroker() *pullBroker {
+	broker := &pullBroker{
+		stopCh:      make(chan struct{}),
+		publishCh:   make(chan pullResult, 1),
+		subscribeCh: make(chan chan pullResult, 1),
+	}
+	go broker.start()
+	return broker
+}
+
+func (p *pullBroker) start() {
+	subscribers := map[chan pullResult]struct{}{}
+	for {
+		select {
+		case <-p.stopCh:
+			return
+		case ch := <-p.subscribeCh:
+			subscribers[ch] = struct{}{}
+		case res := <-p.publishCh:
+			for ch := range subscribers {
+				select {
+				case ch <- res:
+				default:
+				}
+			}
+		}
+	}
+}
+
+func (p *pullBroker) stop() {
+	close(p.stopCh)
+}
+
+func (p *pullBroker) subscribe(ch chan pullResult) {
+	p.subscribeCh <- ch
+}
+
+func (p *pullBroker) publish(res pullResult) {
+	p.publishCh <- res
 }
 
 func newParallelImagePuller(imageService kubecontainer.ImageService, maxParallelImagePulls *int32) imagePuller {
 	if maxParallelImagePulls == nil || *maxParallelImagePulls < 1 {
-		return &parallelImagePuller{imageService, nil}
+		return &parallelImagePuller{imageService, nil, sync.Map{}}
 	}
-	return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls)}
+	return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls), sync.Map{}}
 }
 
-func (pip *parallelImagePuller) pullImage(ctx context.Context, spec kubecontainer.ImageSpec, pullSecrets []v1.Secret, pullChan chan<- pullResult, podSandboxConfig *runtimeapi.PodSandboxConfig) {
+func (pip *parallelImagePuller) pullImage(ctx context.Context, spec kubecontainer.ImageSpec, pullSecrets []v1.Secret, pullChan chan pullResult, podSandboxConfig *runtimeapi.PodSandboxConfig) {
 	go func() {
 		if pip.tokens != nil {
 			pip.tokens <- struct{}{}
 			defer func() { <-pip.tokens }()
 		}
+
+		key := fmt.Sprintf("%s-%s-%s", spec.Image, spec.RuntimeHandler, spec.Annotations)
```

Comment:
If we don't move the broker handling before the token bucket processing at L106-109, won't the goroutine waiting for the container image pull to complete occupy a seat for the maximum parallel image pulls?

Reply:
+1

Reply:
I was thinking of a change similar to this. But I think the best way to do it is to replace the `pip.tokens` with the queue and workers processing. Waiting queue can be iterated over each time one pull finished and weak matching images can be checked if they were pulled already by calling the GetImage.

Some advanced queue prioritization may be done later as well

Reply:
Fixed, we may also want to use a semaphore or similar instead of pip.tokens. I can follow-up on that.